File: DS-Algorithms/Graph/dfs.cpp
#include <iostream>
#include <list>
#include <vector>

using namespace std;

class Graph {
public:
  Graph(int V) : V(V), adj(new list<int>[V]) {}

  void AddEdge(int u, int v) { adj[u].push_back(v); }

  void dfs(int v) {
    vector<bool> marked(V, false);
    list<int> Q;
    dfs_util(v, marked);
  }

private:
  int V;
  list<int> *adj;

  void dfs_util(int v, vector<bool> &marked) {
    marked[v] = true;
    cout << v << " ";
    for (auto it = adj[v].begin(); it != adj[v].end(); ++it) {
      if (!marked[*it]) {
        dfs_util(*it, marked);
      }
    }
  }
};

int main() {

  Graph g(4);
  g.AddEdge(0, 1);
  g.AddEdge(0, 2);
  g.AddEdge(1, 2);
  g.AddEdge(2, 0);
  g.AddEdge(2, 3);
  g.AddEdge(3, 3);

  cout << "dfs: ";
  g.dfs(2);
  cout << "\n";

  return 0;
}

File: 17122019/java-revisited/src/Graph/DFS.java
package Graph;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class DFS {

    private int V;
    private ArrayList<Integer> adj[];

    DFS(int V) {
        this.V = V;
        adj = new ArrayList[V];
        for (int i = 0; i < V; i++) {
            adj[i] = new ArrayList<>();
        }
    }

    void addEdge(int u, int v) {
        adj[u].add(v);
    }

    void dfsUtil(int v, boolean marked[]) {
        marked[v] = true;
        System.out.print(v + " ");
        Iterator<Integer> iterator = adj[v].listIterator();
        while (iterator.hasNext()) {
            v = iterator.next();
            if (!marked[v]) {
                dfsUtil(v, marked);
            }
        }
    }

    void dfs(int v) {
        boolean marked[] = new boolean[V];
        dfsUtil(v, marked);
    }

    public static void main(String[] args) {
        DFS dfs = new DFS(4);
        dfs.addEdge(0, 1);
        dfs.addEdge(0, 2);
        dfs.addEdge(1, 2);
        dfs.addEdge(2, 0);
        dfs.addEdge(2, 3);
        dfs.addEdge(3, 3);
        dfs.dfs(2);
    }
}
